Put stunnel in single kafka image

tombentley opened the pull request:
### Type of change

- Refactoring

### Description

The PR puts `stunnel` in the kafka images, and changes the cluster operator to use this image for the stunnel sidecars. The advantages are:

* Fewer images
* It should avoid the race when kube pulls the images when the sidecar image is downloaded first only to error because the other container hasn't started yet due to slow download.

Note that as for the images used for zookeeper in the single-Kafka change, we configure the CO deployment to use the most recent Kafka image as the default for the stunnel images. It should, however, be fine to run a tls sidecar container that's in a `x-kafka-y` image as a sidecar to a container using a `x-kafka-z` (where y ≠ z); it will just look a bit weird.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

scholzj commented:
> It should avoid the race when kube pulls the images when the sidecar image is downloaded first only to error because the other container hasn't started yet due to slow download.

So, have you tried to test this?

tombentley commented:
> So, have you tried to test this?

No, not yet.

ppatierno commented:
> a tls sidecar container that's in a x-kafka-y image as a sidecar to a container using a x-kafka-z (where y ≠ z)

why something like that should happen?

tombentley commented:
> why something like that should happen?

You've deployed Strimzi 0.x.0 running Kafka 2.y and later upgrade the Kafka to 2.z. If you never specified explicitly the image to use for the sidecar which image should be used both before and after the upgrade? Currently the CO have env vars which configure it with the _single_ defaults to use. Because there is a _single_ default we necessarily end up with the situation where you have a different kafka image in use for the sidecar than for the main container. This is not a problem because it's exactly the same `stunnel` binary in both those images anyway.

Why I highlighted this is precisely because we need to decide whether the defaulting mechanism we have in place is what we want. What I've done so far results in this awkward situation, but it means we've not changed how the CO is configured (the env vars are the same). Changing the CO env vars might help avoid this awkward situation, but comes at the cost of more complex upgrades of the CO for those users who have gone to the trouble to configure it differently from how it comes OOTB.

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityOperator.java:
```diff
@@ -155,6 +157,13 @@ public static EntityOperator fromCrd(Kafka kafkaAssembly) {
             }
+
+            KafkaClusterSpec kafkaClusterSpec = kafkaAssembly.getSpec().getKafka();
+            String tlsSidecarImage = versions.kafkaImage(kafkaClusterSpec.getImage(), kafkaClusterSpec.getVersion());
```
Does this mean we will have to update the EO pod everytime Kafka version changes? Wouldn't it be easier to just use the image for the default version?

tombentley commented:
> I think that the update to the olm is missing again. Or did I just missed it?

That one accidentally ended up in the PR for the operator image, but further changes are needed for the operator image.

tombentley commented:
> Does this mean we will have to update the EO pod everytime Kafka version changes? Wouldn't it be easier to just use the image for the default version?

I've updated the PR to apply a consistent lookup order for the stunnel images, and added tests for that. Could you review that change @scholzj?
